### PR TITLE
Shared testing workflows

### PR DESCRIPTION
**Pull request by AlCalzone**
**PR Checklist:**

-   [x] Provide a meaningful description to this PR or mention which issues this fixes.
-   [ ] ~~Add tests for your change. This includes negative tests (i.e. inputs that need to fail) as well as baseline tests (i.e. how should the directory structure look like?).~~
-   [x] Run the test suite with `npm test`
-   [x] If there are baseline changes, review them and make a separate commit for them with the comment "accept baselines" if they are desired changes
-   [x] Ensure the project builds with `npm run build`
-   [ ] ~~If you added a required option, also add it to the template creation (`.github/create_templates.ts`)~~
-   [x] Add a detailed migration description to `docs/updates` explaining what the user needs to do when manually updating an existing project
-   [x] Add your changes to `CHANGELOG.md` (referencing the migration description and this PR or the issue you fixed)

**Description:**  
With this PR, we make use of GitHub's new composite actions to greatly simplify the testing workflow file.

### CODE REVIEW

**Comment by AlCalzone** [comment]
@UncleSamSwiss I fear this conflicts slightly with #818. Depending on what is merged first, we'll have to rebase.

File: templates/_github/workflows/test-and-release.yml.ts [hunk 1]
```diff
@@ -2,194 +2,117 @@ import type { TemplateFunction } from "../../../src/lib/createAdapter";
...
+        node-version: [${adapterTestVersions.join(", ")}]
+        os: [${adapterTestOS.join(", ")}]
 ${(adapterTestOS.includes("windows-latest") && adapterTestVersions.includes("8.x")) ? (
-	// This is unnecessary but I'm leaving it here in case we need it again.
-	// The else branch will never trigger
-`                exclude:
-                    # Don't test Node.js 8 on Windows. npm is weird here
-                    - os: windows-latest
-                      node-version: 8.x`
+  // This is unnecessary but I'm leaving it here in case we need it again.
```
Comment (UncleSamSwiss) [inline templates/_github/workflows/test-and-release.yml.ts]: I think, we can remove this. If we ever need it again, we can simply look at the git log.

File: templates/_github/workflows/test-and-release.yml.ts [hunk 2]
```diff
@@ -2,194 +2,117 @@ import type { TemplateFunction } from "../../../src/lib/createAdapter";
...
+    steps:
+      - uses: ioBroker/testing-action-adapter@v1
+        with:
+          node-version: \${{ matrix.node-version }}
+          os: \${{ matrix.os }}
+          # Uncomment the following line if your adapter cannot be installed using 'npm ci'
+          # install-command: 'npm install'${needsBuild ? (`
+          build: true`) : ""}
 `) : ""}
 # TODO: To enable automatic npm releases, create a token on npmjs.org
```
Comment (UncleSamSwiss) [inline templates/_github/workflows/test-and-release.yml.ts:73]: Don't worry about incompatibilities with my PR, I will merge your changes.

File: test/baselines/adapter_JS_ES6Class_ESLint_TypeChecking_Spaces_SingleQuotes_Apache-2.0/.github/workflows/test-and-release.yml [hunk 3]
```diff
@@ -3,156 +3,80 @@ name: Test and Release
...
 on:
-    push:
-        branches:
-            - "*"
-        tags:
-            # normal versions
-            - "v[0-9]+.[0-9]+.[0-9]+"
-            # pre-releases
-            - "v[0-9]+.[0-9]+.[0-9]+-**"
+  push:
```
Comment (UncleSamSwiss) [inline test/baselines/adapter_JS_ES6Class_ESLint_TypeChecking_Spaces_SingleQuotes_Apache-2.0/.github/workflows/test-and-release.yml:6]: You switched from 4 to 2 spaces for this yml file. Is this going to cause an issue with prettier or so? Is this consistent with other yml files?

File: docs/updates/20210913_shared_testing_workflows.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+Github recently added support for [composite actions](https://github.blog/changelog/2021-08-25-github-actions-reduce-duplication-with-action-composition/) which finally allow us to provide shared testing workflows:
+
+-   https://github.com/ioBroker/testing-action-check - Package file checks, linting, type checking
+-   https://github.com/ioBroker/testing-action-adapter - Adapter integration tests
+-   https://github.com/ioBroker/testing-action-deploy - Automatic deploy to npm, Github Releases and Sentry
+
+This leads to a greatly simplified testing workflow file. To migrate, we recommend:
+a) creating a new adapter directory with the creator and copying `.github/workflows/test-and-release.yml` to your adapter directory  
+**or**  
+b) using the replay function.
```
Comment (UncleSamSwiss) [inline docs/updates/20210913_shared_testing_workflows.md]: We should probably clarify what the "replay function" is. Most people never heard of this feature of the adapter creator. Perhaps write a sample command.

**Review by UncleSamSwiss — CHANGES_REQUESTED** [review]
Looks good in general. I have some questions and one improvement.

File: test/baselines/adapter_JS_ES6Class_ESLint_TypeChecking_Spaces_SingleQuotes_Apache-2.0/.github/workflows/test-and-release.yml (hunk 3 above)
Comment (AlCalzone) [reply test/baselines/adapter_JS_ES6Class_ESLint_TypeChecking_Spaces_SingleQuotes_Apache-2.0/.github/workflows/test-and-release.yml:6]: We currently exclude the `.yml` files in `.github` from Prettier. I've found 4 spaces to be unnecessarily wide in yml files. The Dependabot auto-merge workflow and Dependabot config already use 2 spaces, the Dependabot auto-merge config uses 4.

File: templates/_github/workflows/test-and-release.yml.ts (hunk 1 above)
Comment (AlCalzone) [reply templates/_github/workflows/test-and-release.yml.ts]: Fine with me

File: docs/updates/20210913_shared_testing_workflows.md [hunk 5]
```diff
@@ -6,10 +6,16 @@ Github recently added support for [composite actions](https://github.blog/change
...
-This leads to a greatly simplified testing workflow file. To migrate, we recommend:
+This leads to a greatly simplified testing workflow file. To migrate, we recommend:  
 a) creating a new adapter directory with the creator and copying `.github/workflows/test-and-release.yml` to your adapter directory  
-**or**  
-b) using the replay function.
+**- or -**  
+b) using the replay function to shorten this process:
+
+```
+npx @iobroker/create-adapter --target=/somewhere/else --replay=/your/adapter/.    create-adapter.json
```
Comment (UncleSamSwiss) [inline docs/updates/20210913_shared_testing_workflows.md]: Is it me or are there 4 spaces where there shouldn't be any?

**Review by UncleSamSwiss — CHANGES_REQUESTED** [review]
One minor thing.

File: docs/updates/20210913_shared_testing_workflows.md (hunk 5 above)
Comment (AlCalzone) [reply docs/updates/20210913_shared_testing_workflows.md]: wtf

**Review by UncleSamSwiss — APPROVED** [review]
:)

**Pull request merged**